src/v/wasm/parser/parser.cc

```diff
@@ -0,0 +1,567 @@
+    }
+}
+
+ss::sstring parse_name(iobuf_parser_base* parser) {
+    auto str_len = leb128::decode<uint32_t>(parser);
+    if (str_len > max_name_length) {
+        throw parse_exception(fmt::format("name too long: {}", str_len));
+    }
+    auto b = parser->read_string(str_len);
+    if (!is_valid_utf8(b)) {
+        throw parse_exception("name invalid utf8");
+    }
+    return {b.begin(), b.end()};
+}
+
+template<size_t kMax>
+std::vector<val_type> parse_signature_types(iobuf_parser_base* parser) {
+    auto vector_size = leb128::decode<uint32_t>(parser);
+    if (vector_size > kMax) {
+        throw module_too_large_exception(
+          fmt::format("too many parameters to function: {}", vector_size));
+    }
+    std::vector<val_type> result_type;
+    result_type.reserve(vector_size);
+    for (uint32_t i = 0; i < vector_size; ++i) {
+        result_type.push_back(parse_val_type(parser));
+    }
+    return result_type;
+}
+
+function_signature parse_signature(iobuf_parser_base* parser) {
+    auto magic = parser->consume_type<uint8_t>();
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+    if (magic != 0x60) {
+        throw parse_exception(
+          fmt::format("function type magic mismatch: {}", magic));
+    }
+    auto parameter_types = parse_signature_types<max_function_params>(parser);
+    auto result_types = parse_signature_types<max_function_results>(parser);
+    return {
+      .parameters = std::move(parameter_types),
+      .results = std::move(result_types)};
+}
+
+declaration::limits parse_limits(iobuf_parser_base* parser) {
+    if (parser->read_bool()) {
+        auto min = leb128::decode<uint32_t>(parser);
+        auto max = leb128::decode<uint32_t>(parser);
+        return {.min = min, .max = max};
+    } else {
+        auto min = leb128::decode<uint32_t>(parser);
+        return {.min = min, .max = std::numeric_limits<uint32_t>::max()};
+    }
+}
+
+declaration::table parse_table_type(iobuf_parser_base* parser) {
+    auto reftype = parse_val_type(parser);
+    if (reftype != val_type::externref && reftype != val_type::funcref) {
+        throw parse_exception(
+          fmt::format("invalid tabletype type: {}", uint8_t(reftype)));
+    }
+    auto limits = parse_limits(parser);
+    return {.reftype = reftype, .limits = limits};
+}
+
+declaration::memory parse_memory_type(iobuf_parser_base* parser) {
+    return {.limits = parse_limits(parser)};
+}
+
+declaration::global parse_global_type(iobuf_parser_base* parser) {
+    auto valtype = parse_val_type(parser);
+    auto mut = parser->read_bool();
+    return {.valtype = valtype, .is_mutable = bool(mut)};
+}
+
+class module_extractor {
+public:
+    explicit module_extractor(iobuf_parser_base* parser)
+      : _parser(parser) {}
+
+    // Parse a WebAssembly binary module.
+    //
+    // This specific function mostly checks the magic bytes and version, then
+    // parses each section of the module.
+    ss::future<module_declarations> parse() {
```

Comment:
> parsed recursively

there is no recursion here FWIW, this parser is not bound by the stack size, and this is supposed to be designed to be easy to parse.

> does this extend to the whole binary spec? As in, if we wanted to parse and validate a whole program, would we need to rework this code significantly (or at all)?

Yes it does extend to the whole spec, I have a version of this that goes farther and does full validation (we don't validate everything is right here) along with parsing instructions. We don't have the need for knowing anything about the instructions, so we can skip that bit for now. Also the instruction space is changing more rapidly than this space.

Reply:
>no recursion

I just mean bounded recursion in parser code e.g. parse_import -> {parse_name, parse_name, parse_X_type} as opposed to some version of a pushdown automaton. 

A better question would have been whether the entire WASM binary format could be parsed with a single byte of lookahead and no backtracking. Sounds like the answer is yes.

Reply:
Ah yes - it is context free so it can be done with pushdown automaton.

Yes no backtracking is needed to parse wasm, a requirement is that it can be parsed in a single linear scan, as it's desirable that you can even do compilation as you parse in a single pass compiler. There are a couple proven implementations of this.

Reply:
Yuup. Pretty slick if you ask me!